### ISSUE TITLE
log_probability returns -inf.

### ISSUE THREAD

**Issue by MARIUSDRAGOS96** [jmschrei/pomegranate#916]
Hello!

I am building an **HMM** from samples like follows:
`    def train(self):
        """
        Train the model using the provided training data set

        :return: None
        """
        seed = 123
        numpy.random.seed(seed)

        sequences = [self.states, self.training_data_set]

        self.model = HiddenMarkovModel.from_samples(
            DiscreteDistribution,
            self.number_of_components,
            sequences,
            n_jobs=10)`
The training data set contains values of energy measured at a certain timestamp:
`0
0
0
0
0
0
64.2
0
0
46.8
70.4
0
0
0
0
0
`
These are the first 16 values from the file. Before building the model, these values are scaled.
I am predicting the next state by building windows of different lengths, for example [0, 0, 0, 0, 0, 0, 1, 0], and adding each other possible state at the end this sequence and retrieve the log probability:
`    def __get_predicted_next_state(self, state_sequence):
        """
        Get the predicted next state based on a given state sequence

        :param state_sequence: state sequence based on which to get the predicted next state
        :type state_sequence: numpy.ndarray

        :return: predicted next state
        :type: int
        """
        maximum_probability = -math.inf
        predicted_next_state = -1

        for state in self.training_states:
            current_state_sequence = numpy.append(state_sequence, state)

            probability = self.model.log_probability(current_state_sequence)

            if probability > maximum_probability:
                maximum_probability = probability
                predicted_next_state = state

        return predicted_next_state`
The sequence that has the biggest log probability gives the predicted value after the given window.
However, for some given sequences, the log_probability method always returns **-inf**.
What could be the issue?

Thank you!

**Comment by jmschrei** [comment]
If you're using DiscreteDistributions you will get `-inf` whenever you encounter a symbol you haven't seen in the training set. If your data is numeric, like it seems to be, you should try using something like an ExponentialDistribution or NormalDistribution.

**Comment by MARIUSDRAGOS96** [comment]
Thanks for the quick reply.
Using the **ExponentialDistribution**, the log_probability returns always **0** as being the next predicted state.
Using the **NormalDistribution**, the log_probability returns always **-inf**.

**Comment by jmschrei** [comment]
I think you're formulating the problem incorrectly. If you have a sequence of length n and a state sequence of length n+1, the model will probably have difficulty finding a way to map those n observations onto n+1 states when you call `model.log_probability`. 

I'm not sure what exactly the problem you're trying to solve is, but here are two potential problems and solutions:

You know what the n+1th observation is and want to determine the n+1th state: You can run normal inference algorithms to get this, e.g., using `predict_proba`.  

You know the nth observation and nth state but want to guess what the n+1th state is: In this case, you can just look at the row in the transition matrix `model.dense_transition_matrix` for the state you're currently in. Given that you know nothing about the emissions, the emission probability distributions don't matter.

Let me know if I missed something.

**Comment by MARIUSDRAGOS96** [comment]
Sorry, I am a beginner in this topic.
Let me describe the requirements.
So, I have multiple **TXT** files having a level of energy consumption/production, depending on the file, on each line.
What I want to achieve is to train an HMM model using one of these files, or a portion of it, and based on that model to be able to predict the next value that comes after a certain sequence of values.
For example:
Train the model with `0
0
0
0
0
0
64.2
0
0
46.8
70.4
0
0
0
0
0
0
0
0
0` and let the model then guess what value comes after this sequence of values `0
0
0
0
68
257.2
146.4
0
0
0
0`
Could you give some directions on how can I retrieve some probabilities or how to predict that next value?

**Comment by jmschrei** [comment]
That doesn't sound like a problem an HMM would be good for. HMMs are good at tagging each observation with an underlying hidden state that is not already known. It sounds like you're looking for an autoregressive model. The simplest thing you can do is train a linear regression model on chunks of k observations and have it predict the k+1th observation, though naturally you can use as complicated a supervised machine learning model as you'd like, including an RNN.
